### PR TITLE
nerfplayer with ngp and nerfacto

### PR DESCRIPTION
**Pull request by lsongx**
NeRFPlayer implementation with `instant-ngp-bounded` and `nerfacto` model

Examples:
With `nerfacto` as the base model
nerfacto

With `instant-ngp-bounded` as the base model
instantngp

### CODE REVIEW

**Comment by machenmusik** [comment]
Nice!  I've been able to reproduce results, but nerfacc wants to set up CUDA again every time... which takes quite a while.  Not sure whether that is related to changes here, or separate issue.  Do others see the same behavior?

File: nerfstudio/fields/nerfplayer_nerfacto_field.py [hunk 1]
```diff
@@ -0,0 +1,409 @@
...
+
+
+class TemporalHashMLPDensityField(Field):
+    """A lightweight temporal density field module.
+
+    Args:
+        aabb: parameters of scene aabb bounds
+        num_layers: number of hidden layers
+        hidden_dim: dimension of hidden layers
+        spatial_distortion: spatial distortion module
```
Comment (tancik) [inline nerfstudio/fields/nerfplayer_nerfacto_field.py:66]: Add missing arg docs

**Review by tancik — COMMENTED** [review]
Looks great!

**Comment by machenmusik** [comment]
> Thank you @machenmusik ! It needs to compile `nerfstudio\field_components\cuda` for the first run, which can take up to 20mins. (`MAX_JOBS` env for ninja may help speed up this). But after compiling, I do not need to compile later. Would you mind providing a little more details about 'setting up CUDA again every time'?

For example, I trained the mochi-high-five dataset (`ns-train nerfplayer-nerfacto ...`) to completion; before it started training, I got the usual `Setting up CUDA (This may take a few minutes the first time)` from nerfacc.

But now that I am loading that previously trained checkpoint (`ns-train nerfplayer-nerfacto --load-dir ...`), I am seeing the CUDA setup message and delay again, which I would not expect.

I do notice that I see it from NerfAcc, and then a similar message again
`nerfstudio field components: Setting up CUDA (This may take a few minutes the first time)` - 
I don't remember seeing `nerfstudio field components` previously, but not sure if that is now expected.

**Comment by lsongx** [comment]
> For example, I trained the mochi-high-five dataset (`ns-train nerfplayer-nerfacto ...`) to completion on Windows 10 machine; before it started training, I got the usual `Setting up CUDA (This may take a few minutes the first time)` from nerfacc.
> 
> But now that I am loading that previously trained checkpoint (`ns-train nerfplayer-nerfacto --load-dir ...`), I am seeing the CUDA setup message and delay again, which I would not expect.
> 
> I do notice that I see it from NerfAcc, and then a similar message again `nerfstudio field components: Setting up CUDA (This may take a few minutes the first time)` - I don't remember seeing `nerfstudio field components` previously, but not sure if that is now expected.

Yes, this compilation is expected. Some more CUDA kernels are needed to run nerfplayer -- implementation can be found at `nerfstudio/field_components/cuda`. If this message appears only once on a new machine, I think that is fine.

**Comment by machenmusik** [comment]
> Yes, this compilation is expected. Some more CUDA kernels are needed to run nerfplayer -- implementation can be found at `nerfstudio/field_components/cuda`. If this message appears only once on a new machine, I think that is fine.

Sure, but the issue is that it was required the second time through `ns-train nerfplayer-nerfacto` as well (which is still processing now).

**Comment by lsongx** [comment]
> Sure, but the issue is that it was required the second time through `ns-train nerfplayer-nerfacto` as well (which is still processing now).

I see. Thanks for helping debug this! I think it is related to the `.so` file, which may not exist on Windows. https://github.com/nerfstudio-project/nerfstudio/blob/828d6bc893683042b3763247b65b3adeb5d5ea7e/nerfstudio/field_components/cuda/_backend.py#L30
I updated a new version. Would you mind giving it a try? Thanks again @machenmusik

**Comment by machenmusik** [comment]
> > Sure, but the issue is that it was required the second time through `ns-train nerfplayer-nerfacto` as well (which is still processing now).
> 
> I see. Thanks for helping debug this! I think it is related to the `.so` file, which may not exist on Windows.
> 
> https://github.com/nerfstudio-project/nerfstudio/blob/828d6bc893683042b3763247b65b3adeb5d5ea7e/nerfstudio/field_components/cuda/_backend.py#L30
> 
> 
> Trying to solve this now...

Thanks.  I am trying the latest, still not done with first CUDA compile yet.

**Comment by machenmusik** [comment]
This worked:
```
_C = None
if cuda_toolkit_available():
    if os.listdir(BUILD_DIR) != []:
        # If the build exists, we assume the extension has been built
        # and we can load it.
        print("nerfstudio field components: CUDA set up, loading (should be quick)")
        _C = load(
            name=NAME,
            sources=glob.glob(os.path.join(PATH, "csrc/*.cu")),
            extra_cflags=["-O3", "-std=c++14"],
            extra_cuda_cflags=["-O3", "-std=c++14"],
            extra_include_paths=[],
        )
    else:
        # Build from scratch. Remove the build directory just to be safe: pytorch jit might stuck
        # if the build directory exists.
        shutil.rmtree(BUILD_DIR)
        print("nerfstudio field components: Setting up CUDA (This may take a few minutes the first time)")
        _C = load(
            name=NAME,
            sources=glob.glob(os.path.join(PATH, "csrc/*.cu")),
            extra_cflags=["-O3", "-std=c++14"],
            extra_cuda_cflags=["-O3", "-std=c++14"],
            extra_include_paths=[],
        )
        print("nerfstudio field components: Setting up CUDA finished")
else:
    print("nerfstudio field components: No CUDA toolkit found. Some models may fail.")
```

Also https://github.com/KAIR-BAIR/nerfacc/pull/150

**Comment by machenmusik** [comment]
Pushed the working change -- note that https://github.com/KAIR-BAIR/nerfacc/pull/150 is also needed or else nerfacc 0.3.x has same problem.  Does this still work on Linux?

**Comment by lsongx** [comment]
> Pushed the working change -- note that [KAIR-BAIR/nerfacc#150](https://github.com/KAIR-BAIR/nerfacc/pull/150) is also needed or else nerfacc 0.3.x has same problem. Does this still work on Linux?

Yes. This works. Thanks!
If a user cancels the compilation in the middle, the folder is not empty. But loading a non-empty and unfinished build may have some issues (from this comment)
https://github.com/KAIR-BAIR/nerfacc/blob/58d6b4c46dfee36ab4b565ab0655b1a27ff4859f/nerfacc/cuda/_backend.py#L78

**Comment by machenmusik** [comment]
> Yes. This works. Thanks! If a user cancels the compilation in the middle, the folder is not empty. But loading a non-empty and unfinished build may have some issues (from this comment) https://github.com/KAIR-BAIR/nerfacc/blob/58d6b4c46dfee36ab4b565ab0655b1a27ff4859f/nerfacc/cuda/_backend.py#L78

Library suffixes are generally .so, .dll, or .a - right?  Maybe checking for those would work for both Linux and Windows (and perhaps Mac as well).

**Comment by lsongx** [comment]
> Library suffixes are generally .so, .dll, or .a - right? Maybe checking for those would work for both Linux and Windows (and perhaps Mac as well).

Sounds good. But I cannot find an example of that. Do you know some examples so that we can learn from them?
Or we can simply ignore the suffix, like
```
if NAME in [p.stem for p in Path(BUILD_DIR).glob("*")]:
```

I see that most of the libs do not use the jit load. They move the build to the setup step (i.e., `pip install` step), and load it as a module.  Not sure if this will be good practice for nerfstudio and nerfacc.

**Comment by machenmusik** [comment]
> Sounds good. But I cannot find an example of that. Do you know some examples so that we can learn from them? 

I was thinking simply to do logical OR of exists calls; they should be lightning quick.

> I see that most of the libs do not use the jit load. They move the build to the setup step (i.e., `pip install` step), and load it as a module. Not sure if this will be good practice for nerfstudio and nerfacc.

Right, the install would then take a rather long time even if the user never utilized those functions.

**Comment by lsongx** [comment]
> I was thinking simply to do logical OR of exists calls; they should be lightning quick.

Agree. I think we may wait and see if there are enough people from different platforms willing to help test it.
Besides, NerfAcc @liruilong940607  may also provide a better solution. We can come back to this later.
Do you think currently the directory check solution can get merged? After searching for a while, the stuck problem brought by non-empty build dir looks not common.

**Comment by machenmusik** [comment]
Note: this did NOT work (and would have rebuilt for many minutes if I hadn't commented out the rmtree)
```
if os.path.exists(os.path.join(BUILD_DIR, f"{NAME}.so")) or os.path.exists(os.path.join(BUILD_DIR, f"{NAME}.dll")) or os.path.exists(os.path.join(BUILD_DIR, f"{NAME}.a")):
```
I would merge with the directory check.

**Review by machenmusik — APPROVED** [review]
Thanks for contributing and refining!  @tancik any last requests?

**Comment by tancik** [comment]
> I see that most of the libs do not use the jit load. They move the build to the setup step (i.e., pip install step), and load it as a module. Not sure if this will be good practice for nerfstudio and nerfacc.

I think this is really what we should be doing. Since this is the first cuda code in this repo, maybe we should do this now.

**Comment by lsongx** [comment]
For the sometimes unwanted time-consuming build (they do not use it),  what about providing an extra step for `pip install`? 

The default behavior of `pip install` remains the same (no compilation). Then if someone wants to use it, we first throw an error message, telling the user to run another `pip install [compile setup]`.

**Comment by tancik** [comment]
yea, let's start with that for now, it could be `pip install .[cuda]`

**Comment by machenmusik** [comment]
The only concern I have is that the CUDA compile times are egregiously long, and people should be warned that the extra install will be lengthy.  It may also be worth trying to mention that second install step after doing the initial one for those that want to get it out of the way, and allowing a way to ask for all the compiles to be triggered up front.

If nerfacc doesn't have the same behavior (so still stalls) or requires a separate additional install step, that could get tedious.

**Comment by tancik** [comment]
> The default behavior of pip install remains the same (no compilation). Then if someone wants to use it, we first throw an error message, telling the user to run another pip install [compile setup].

I think this solution of yours solves the issue of it taking a while, since people will only run it if necessary

`Nerfacc` is trickier, we can figure out what to do with that later.

**Comment by machenmusik** [comment]
> > The default behavior of pip install remains the same (no compilation). Then if someone wants to use it, we first throw an error message, telling the user to run another pip install [compile setup].
> 
> I think this solution of yours solves the issue of it taking a while, since people will only run it if necessary
> 
> `Nerfacc` is trickier, we can figure out what to do with that later.

Trickier perhaps, but if we want to avoid JIT stalls for more deterministic behavior, there seems little point in doing so for field components but still stalling on nerfacc's CUDA JIT.  That is actually worse IMO since you have to do two separate CUDA compiles, one of which is unexpected.

**Comment by tancik** [comment]
> Trickier perhaps, but if we want to avoid JIT stalls for more deterministic behavior, there seems little point in doing so for field components but still stalling on nerfacc's CUDA JIT. That is actually worse IMO since you have to do two separate CUDA compiles, one of which is unexpected.

I agree. I've got to chat with @liruilong940607 to get his input first.

**Comment by machenmusik** [comment]
Great.  For pytorch cuda, IIRC we download (pre-built) wheels, perhaps we can do similar for nerfacc and nerfstudio...

**Comment by tancik** [comment]
> Great. For pytorch cuda, IIRC we download (pre-built) wheels, perhaps we can do similar for nerfacc and nerfstudio...

That would be cool, it surpasses my knowledge on python deployment. How easy is that to do in practice? Also does pypi support multiple wheels?

**Comment by machenmusik** [comment]
I am not terribly familiar with the build mechanics either - but it does seem like the right approach.  Volunteers welcome...

**Comment by machenmusik** [comment]
https://discuss.python.org/t/what-to-do-about-gpus-and-the-built-distributions-that-support-them/7125

**Comment by machenmusik** [comment]
From some light research, making wheels available for various permutations seems a challenge;
at the moment I would suggest restoring the behavior we had with nerfacc 0.2.x (with the appropriate change to nerfacc)
https://github.com/nerfstudio-project/nerfstudio/pull/1277#issuecomment-1404215735
until there is a better solution that does not force lengthy CUDA recompilation every time on Windows.

**Comment by machenmusik** [comment]
(It seems unfortunate that we're holding this up on the CUDA JIT question, when the nerfplayer functionality seems to work well.)

**Comment by tancik** [comment]
> (It seems unfortunate that we're holding this up on the CUDA JIT question, when the nerfplayer functionality seems to work well.)

True, is the cuda in a good state now (ie the original issue you brought up @machenmusik is resolved). If so I'm fine with merging and addressing larger cuda changes later.

**Comment by machenmusik** [comment]
@tancik yes the Windows CUDA recompilation is fixed by combination of https://github.com/nerfstudio-project/nerfstudio/pull/1277/commits/59cb8d2ca49fa500bee331337da7e2dfcd26d49a in this PR, and also https://github.com/KAIR-BAIR/nerfacc/pull/150 which is merged but may need a release and dependency bump.

**Comment by tancik** [comment]
Ok, @lsongx feel free to merge when ready.

**Comment by lsongx** [comment]
Thanks, merged! Feel free to mention me if there are any bug reports or feature requests or ... related to these PRs. Will be happy to contribute more, including future CUDA changes.

**Comment by tancik** [comment]
Sounds good, thanks for adding this @lsongx!

**Comment by machenmusik** [comment]
Nice work @lsongx !
@tancik do we need to bump nerfacc version separately?

**Comment by tancik** [comment]
@machenmusik the nerfacc version was just bumped. There are also not nerfacc wheels https://github.com/KAIR-BAIR/nerfacc#installation

**Pull request merged**